app/ethermint.go

```diff
@@ -124,122 +140,184 @@ type EthermintApp struct {
+	app.subspaces[distr.ModuleName] = app.ParamsKeeper.Subspace(distr.DefaultParamspace)
+	app.subspaces[slashing.ModuleName] = app.ParamsKeeper.Subspace(slashing.DefaultParamspace)
+	app.subspaces[gov.ModuleName] = app.ParamsKeeper.Subspace(gov.DefaultParamspace).WithKeyTable(gov.ParamKeyTable())
+	app.subspaces[crisis.ModuleName] = app.ParamsKeeper.Subspace(crisis.DefaultParamspace)
+	app.subspaces[evidence.ModuleName] = app.ParamsKeeper.Subspace(evidence.DefaultParamspace)
 
 	// add keepers
-	app.accountKeeper = auth.NewAccountKeeper(app.cdc, keys[auth.StoreKey], authSubspace, eminttypes.ProtoBaseAccount)
-	app.bankKeeper = bank.NewBaseKeeper(app.accountKeeper, bankSubspace, bank.DefaultCodespace, app.ModuleAccountAddrs())
-	app.supplyKeeper = supply.NewKeeper(app.cdc, keys[supply.StoreKey], app.accountKeeper, app.bankKeeper, maccPerms)
-	stakingKeeper := staking.NewKeeper(app.cdc, keys[staking.StoreKey],
-		app.supplyKeeper, stakingSubspace, staking.DefaultCodespace)
-	app.mintKeeper = mint.NewKeeper(app.cdc, keys[mint.StoreKey], mintSubspace, &stakingKeeper, app.supplyKeeper, auth.FeeCollectorName)
-	app.distrKeeper = distr.NewKeeper(app.cdc, keys[distr.StoreKey], distrSubspace, &stakingKeeper,
-		app.supplyKeeper, distr.DefaultCodespace, auth.FeeCollectorName, app.ModuleAccountAddrs())
-	app.slashingKeeper = slashing.NewKeeper(app.cdc, keys[slashing.StoreKey], &stakingKeeper,
-		slashingSubspace, slashing.DefaultCodespace)
-	app.crisisKeeper = crisis.NewKeeper(crisisSubspace, invCheckPeriod, app.supplyKeeper, auth.FeeCollectorName)
-	app.evmKeeper = evm.NewKeeper(app.accountKeeper, keys[evmtypes.EvmStoreKey], keys[evmtypes.EvmCodeKey], blockKey, cdc)
+	app.AccountKeeper = auth.NewAccountKeeper(
+		appCodec, keys[auth.StoreKey], app.subspaces[auth.ModuleName], auth.ProtoBaseAccount,
+	)
+	app.BankKeeper = bank.NewBaseKeeper(
+		app.AccountKeeper, app.subspaces[bank.ModuleName], app.BlacklistedAccAddrs(),
+	)
+	app.SupplyKeeper = supply.NewKeeper(
+		appCodec, keys[supply.StoreKey], app.AccountKeeper, app.BankKeeper, maccPerms,
+	)
+	stakingKeeper := staking.NewKeeper(
+		appCodec, keys[staking.StoreKey], app.SupplyKeeper, app.subspaces[staking.ModuleName],
+	)
+	app.MintKeeper = mint.NewKeeper(
+		appCodec, keys[mint.StoreKey], app.subspaces[mint.ModuleName], &stakingKeeper,
+		app.SupplyKeeper, auth.FeeCollectorName,
+	)
+	app.DistrKeeper = distr.NewKeeper(
+		appCodec, keys[distr.StoreKey], app.subspaces[distr.ModuleName], &stakingKeeper,
+		app.SupplyKeeper, auth.FeeCollectorName, app.ModuleAccountAddrs(),
+	)
+	app.SlashingKeeper = slashing.NewKeeper(
+		appCodec, keys[slashing.StoreKey], &stakingKeeper, app.subspaces[slashing.ModuleName],
+	)
+	app.CrisisKeeper = crisis.NewKeeper(
+		app.subspaces[crisis.ModuleName], invCheckPeriod, app.SupplyKeeper, auth.FeeCollectorName,
+	)
+	app.EvmKeeper = evm.NewKeeper(
+		appCodec, keys[evm.BlockKey], keys[evm.CodeKey], keys[evm.StoreKey], app.AccountKeeper,
```

Comment:
The BlockKey reference doesn't exist because it isn't in keys, here is the failure on run (iirc I had blockkey defined seperately because I was using a different db type, but this could probably be improved anyway to have nil handling since this really should be optional anyway):

```
E[2020-03-11|09:18:19.526] CONSENSUS FAILURE!!!                         module=consensus err="interface conversion: interface is nil, not types.KVStore" stack="goroutine 123 [running]:\nruntime/debug.Stack(0xc000d37e70, 0x4e1d480, 0xc000d348a0)\n\t/usr/local/Cellar/go/1.13.5/libexec/src/runtime/debug/stack.go:24 +0x9d\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2(0xc000b6c700, 0x51361a8)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:613 +0x57\npanic(0x4e1d480, 0xc000d348a0)\n\t/usr/local/Cellar/go/1.13.5/libexec/src/runtime/panic.go:679 +0x1b2\ngithub.com/cosmos/cosmos-sdk/store/cachemulti.Store.GetKVStore(...)\n\t/Users/austinabell/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.38.1/store/cachemulti/store.go:153\ngithub.com/cosmos/cosmos-sdk/types.Context.KVStore(0x5371320, 0xc00003c0b8, 0x5385fa0, 0xc00031c380, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.38.1/types/context.go:212 +0x89\ngithub.com/cosmos/ethermint/x/evm/keeper.(*Keeper).SetBlockBloomMapping(0xc000d38da8, 0x5371320, 0xc00003c0b8, 0x5385fa0, 0xc00031c380, 0xa, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/development/github.com/chainsafe/ethermint/x/evm/keeper/keeper.go:92 +0x75\ngithub.com/cosmos/ethermint/x/evm.AppModule.BeginBlock(0xc000109c70, 0x53601a0, 0x0, 0xc00009c800, 0xc00046af58, 0xc000329680, 0x5371320, 0xc00003c0b8, 0x5385fa0, 0xc00031c380, ...)\n\t/Users/austinabell/development/github.com/chainsafe/ethermint/x/evm/module.go:113 +0x17d\ngithub.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(0xc000105030, 0x5371320, 0xc00003c0b8, 0x5385fa0, 0xc00031c380, 0xa, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.38.1/types/module/module.go:297 +0x1ca\ngithub.com/cosmos/ethermint/app.(*EthermintApp).BeginBlocker(...)\n\t/Users/austinabell/development/github.com/chainsafe/ethermint/app/ethermint.go:324\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).BeginBlock(0xc000aa92c0, 0xc0004842a0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.38.1/baseapp/abci.go:136 +0x469\ngithub.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(0xc0003332c0, 0xc0004842a0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/abci/client/local_client.go:231 +0x101\ngithub.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(0xc000e6b920, 0xc0004842a0, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/proxy/app_conn.go:69 +0x6b\ngithub.com/tendermint/tendermint/state.execBlockOnProxyApp(0x5371d20, 0xc000c214a0, 0x537f160, 0xc000e6b920, 0xc000118fc0, 0x53880a0, 0xc00032a0c0, 0x6, 0x5c842b8, 0x1)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/state/execution.go:280 +0x3e1\ngithub.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(0xc000ac79d0, 0xa, 0x0, 0xc000ae44b0, 0x6, 0x5c842b8, 0x1, 0x0, 0x0, 0x0, ...)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/state/execution.go:131 +0x17a\ngithub.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0xc000b6c700, 0x1)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1431 +0x8f5\ngithub.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0xc000b6c700, 0x1)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1350 +0x383\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit.func1(0xc000b6c700, 0x0, 0x1)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1285 +0x90\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit(0xc000b6c700, 0x1, 0x0)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1322 +0x61a\ngithub.com/tendermint/tendermint/consensus.(*State).addVote(0xc000b6c700, 0xc000ec0dc0, 0x0, 0x0, 0xc000cf5ac0, 0x4030c51, 0x51375f0)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1819 +0xa39\ngithub.com/tendermint/tendermint/consensus.(*State).tryAddVote(0xc000b6c700, 0xc000ec0dc0, 0x0, 0x0, 0x8, 0xc000cf5ad0, 0x40b74ed)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:1642 +0x59\ngithub.com/tendermint/tendermint/consensus.(*State).handleMsg(0xc000b6c700, 0x534d9a0, 0xc00032aac0, 0x0, 0x0)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:709 +0x252\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc000b6c700, 0x0)\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:660 +0x579\ncreated by github.com/tendermint/tendermint/consensus.(*State).OnStart\n\t/Users/austinabell/go/pkg/mod/github.com/tendermint/tendermint@v0.33.0/consensus/state.go:335 +0x13a\n"
^CE[2020-03-11|09:18:26.238] Not stopping BlockPool -- have not been started yet module=blockchain impl=BlockPool
E[2020-03-11|09:18:26.238] Stopped accept routine, as transport is closed module=p2p numPeers=0
```

Reply:
good catch!